Module 3: Introduction to NN Frameworks

Problem Statement: This assignment challenges students to construct and train a neural network to accurately classify handwritten digits from the MNIST dataset. The goal is to navigate through the preprocessing of image data, select an appropriate neural network architecture, and apply optimization techniques to achieve high classification accuracy

Introduction: The MNIST dataset is a large database of handwritten digits commonly used for training various image processing systems. The dataset contains 60,000 training images and 10,000 testing images, each of which is a 28x28 pixel grayscale image. The goal is to build a neural network model that can accurately classify these images into one of the ten digit classes (0 through 9).

Objectives: Data Preprocessing: Prepare the MNIST dataset for training and testing the neural network model. This includes normalization of image pixel values and converting the output labels to a one-hot encoded format.

Model Development: Create a neural network model capable of classifying 28x28 pixel grayscale images of handwritten digits. The model should have an input layer, one or more hidden layers, and an output layer.

Training: Train the neural network model using the preprocessed training data. Utilize categorical cross-entropy as the loss function, stochastic gradient descent as the optimizer, and track accuracy as a performance metric.

Evaluation: Evaluate the performance of the trained model using the testing dataset, ensuring that the model generalizes well to unseen data.

Prediction: Implement functionality to make predictions on single images or batches of images, returning the digit class with the highest probability.

Analysis: Analyze the results, understand the model's performance, and explore ways to improve it if necessary

Challenges and Considerations:

Class Imbalance: Ensure that the dataset does not have a significant class imbalance that could negatively affect the model’s performance. If imbalance is present, strategies to mitigate its effects should be considered.

Model Complexity: Choose an appropriate model complexity that is capable of learning from the data without overfitting. This may involve experimenting with different numbers of layers and units, as well as regularization techniques.

Activation Functions: Use appropriate activation functions in the hidden layers (ReLU is commonly used for image data) and softmax activation in the output layer to convert logits to probabilities.

Evaluation Metrics: Focus on accuracy as the primary evaluation metric, but also be open to exploring other metrics that could provide additional insights into the model’s performance.

Hyperparameter Tuning: Experiment with different hyperparameters such as learning rate, batch size, and number of epochs to optimize the model’s performance.

Predictions and Interpretability: Provide clear and interpretable predictions, ensuring that the output probabilities are well-calibrated and meaningful.

Load and Preprocess Data

In [3]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical


In [5]:
# 1. Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

print(f"Original training images shape: {x_train.shape}")
print(f"Original testing images shape: {x_test.shape}")
print(f"Original training labels shape: {y_train.shape}")
print(f"Original testing labels shape: {y_test.shape}")

# 2. Normalize pixel values
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# 3. Reshape the image data
# For a simple dense network, images are flattened from 28x28 to 784
num_pixels = x_train.shape[1] * x_train.shape[2]
x_train = x_train.reshape(x_train.shape[0], num_pixels)
x_test = x_test.reshape(x_test.shape[0], num_pixels)

print(f"Reshaped training images shape: {x_train.shape}")
print(f"Reshaped testing images shape: {x_test.shape}")

# 4. Convert labels to one-hot encoded format
num_classes = 10
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)

print(f"One-hot encoded training labels shape: {y_train.shape}")
print(f"One-hot encoded testing labels shape: {y_test.shape}")

print("MNIST dataset loaded, normalized, reshaped, and labels one-hot encoded successfully.")

Original training images shape: (60000, 28, 28)
Original testing images shape: (10000, 28, 28)
Original training labels shape: (60000,)
Original testing labels shape: (10000,)
Reshaped training images shape: (60000, 784)
Reshaped testing images shape: (10000, 784)
One-hot encoded training labels shape: (60000, 10)
One-hot encoded testing labels shape: (10000, 10)
MNIST dataset loaded, normalized, reshaped, and labels one-hot encoded successfully.


Define Neural Network Model

In [8]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# 1. Instantiate a Sequential model
model = Sequential()

# 2. Add a Dense layer as the input layer
# num_pixels was defined in the previous step as 784
model.add(Dense(128, activation='relu', input_shape=(num_pixels,)))

# 3. Add at least one more Dense hidden layer
model.add(Dense(64, activation='relu'))

# 4. Add a final Dense output layer
# num_classes was defined in the previous step as 10
model.add(Dense(num_classes, activation='softmax'))

# Print the model summary to confirm its structure
print("Neural Network Model Summary:")
model.summary()

print("Neural network model constructed successfully.")

C:\Users\Durga\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Neural Network Model Summary:


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 109,386 (427.29 KB)

 Trainable params: 109,386 (427.29 KB)

 Non-trainable params: 0 (0.00 B)

Neural network model constructed successfully.


In [10]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input

# 1. Instantiate a Sequential model
model = Sequential()

# Add an Input layer explicitly to resolve the UserWarning
model.add(Input(shape=(num_pixels,)))

# 2. Add a Dense layer as the input layer (now implicitly connected to Input layer)
model.add(Dense(128, activation='relu'))

# 3. Add at least one more Dense hidden layer
model.add(Dense(64, activation='relu'))

# 4. Add a final Dense output layer
# num_classes was defined in the previous step as 10
model.add(Dense(num_classes, activation='softmax'))

# Print the model summary to confirm its structure
print("Neural Network Model Summary:")
model.summary()

print("Neural network model constructed successfully.")

Neural Network Model Summary:


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 109,386 (427.29 KB)

 Trainable params: 109,386 (427.29 KB)

 Non-trainable params: 0 (0.00 B)

Neural network model constructed successfully.


Compile Neural Network Model

In [12]:
from tensorflow.keras.optimizers import Adam

# Compile the model
model.compile(
    loss='categorical_crossentropy',
    optimizer=Adam(), # Using Adam optimizer as it's a common and effective choice
    metrics=['accuracy']
)

print("Neural network model compiled successfully with Adam optimizer, categorical_crossentropy loss, and accuracy metric.")

Neural network model compiled successfully with Adam optimizer, categorical_crossentropy loss, and accuracy metric.


Train Neural Network Model

In [14]:
print("Training the neural network model...")

# Define training hyperparameters
epochs = 10
batch_size = 128

# Train the model
history = model.fit(
    x_train,
    y_train,
    epochs=epochs,
    batch_size=batch_size,
    validation_data=(x_test, y_test),
    verbose=1
)

print(f"Neural network model trained for {epochs} epochs with a batch size of {batch_size}.")

Training the neural network model...
Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 12ms/step - accuracy: 0.8271 - loss: 0.6234 - val_accuracy: 0.9551 - val_loss: 0.1527
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.9578 - loss: 0.1423 - val_accuracy: 0.9669 - val_loss: 0.1110
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 10ms/step - accuracy: 0.9715 - loss: 0.0971 - val_accuracy: 0.9713 - val_loss: 0.0921
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 6s 12ms/step - accuracy: 0.9801 - loss: 0.0691 - val_accuracy: 0.9722 - val_loss: 0.0874
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 10ms/step - accuracy: 0.9840 - loss: 0.0540 - val_accuracy: 0.9769 - val_loss: 0.0718
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.9873 - loss: 0.0425 - val_accuracy: 0.9768 - val_loss: 0.0750
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 10ms/step - accuracy: 0.9893 - loss: 0.0339 - val_accuracy: 0.9778 - val_loss: 0.0735
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 5s 10ms/step - accurac

Evaluate Neural Network Model

In [16]:
print("Evaluating the neural network model on the test data...")

loss, accuracy = model.evaluate(x_test, y_test, verbose=0)

print(f"Test Loss: {loss:.4f}")
print(f"Test Accuracy: {accuracy:.4f}")
print("Model evaluation complete.")

Evaluating the neural network model on the test data...
Test Loss: 0.0746
Test Accuracy: 0.9799
Model evaluation complete.


Implement Prediction Functionality

In [18]:
import numpy as np

print("Implementing prediction functionality...")

# 1. Select a few sample images from x_test
# Let's pick the first 5 images for demonstration
sample_indices = range(5)
sample_images = x_test[sample_indices]
sample_true_labels = y_test[sample_indices]

# 3. Use the trained model to make predictions on these sample images
predictions = model.predict(sample_images)

# 4. Interpret the model's output to determine the predicted digit class
predicted_classes = np.argmax(predictions, axis=1)
true_classes = np.argmax(sample_true_labels, axis=1)

# 5. Print the predicted digit along with the true digit for each sample
print("\nPredictions for sample images:")
for i in range(len(sample_indices)):
    print(f"Sample {i+1}: True Digit: {true_classes[i]}, Predicted Digit: {predicted_classes[i]}, Probabilities: {predictions[i]}")

print("Prediction functionality implemented and tested on sample images.")

Implementing prediction functionality...
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 443ms/step

Predictions for sample images:
Sample 1: True Digit: 7, Predicted Digit: 7, Probabilities: [2.36018693e-09 4.01109279e-09 5.53112649e-08 1.12876005e-05
 1.29223563e-12 3.77104792e-10 5.33771595e-13 9.99988317e-01
 2.25833574e-08 2.24490748e-07]
Sample 2: True Digit: 2, Predicted Digit: 2, Probabilities: [5.3544908e-11 9.8994735e-07 9.9999893e-01 3.1192393e-08 1.1941983e-13
 1.4318603e-13 1.5943295e-10 3.5288350e-12 2.3033388e-10 3.0802135e-15]
Sample 3: True Digit: 1, Predicted Digit: 1, Probabilities: [5.20335277e-08 9.99938965e-01 8.88272643e-06 1.02513695e-07
 4.32864499e-06 4.65542769e-07 9.81107678e-06 1.86827856e-05
 1.86461712e-05 1.62553597e-08]
Sample 4: True Digit: 0, Predicted Digit: 0, Probabilities: [9.9999785e-01 5.7220908e-09 5.0204363e-07 1.5809525e-10 7.5047613e-10
 2.4999538e-09 1.1860689e-06 3.0635576e-08 2.8995681e-11 4.4955800e-07]
Sample 5: True Digit: 4, Predicted Digit: 4, Probabili

Analyze and Improve Model

1.The MNIST dataset was successfully loaded and preprocessed. Image pixel values were normalized to a [0,1] range, and 28x28 images were reshaped into 1D arrays of 784 pixels. 
Labels were converted to a 10-class one-hot encoded format
2.A neural network model was constructed with an input layer, two dense hidden layers (128 and 64 units, both with ReLU activation), and an output layer (10 units with softmax activation). The model comprised 109,386 trainable parameters.
3.The model was compiled using categorical_crossentropy as the loss function, the Adam optimizer, and accuracy as the metric.
4.
After training for 10 epochs with a batch size of 128, the model achieved a final training accuracy o 0.99476 and a validation accuracy of 0.9743.5.
The model was evaluated on the test dataset, achieving a Test Loss o0.074625 and a Test Accuracy o0.97994

The model exhibits strong performance, achieving a test accuracy of approximately 97.99%. This suggests it generalizes well to unseen data.3.